### PR TITLE
CAMEL-6840 make it possible grouped throttling

### PR DESCRIPTION
**Pull request by onderson**

### CODE REVIEW

**Comment by oscerd** [comment]
LGTM @onderson, you can wait for @davsclaus or merge directly :-)

File: camel-core/src/main/java/org/apache/camel/model/ProcessorDefinition.java [hunk 1]
```diff
@@ -2284,6 +2284,48 @@ public ThrottleDefinition throttle(Expression maximumRequestCount) {
...
+     * or that we don't exceed an agreed SLA with some external service.
+     * Here another parameter correlationExpressionKey is introduced for the functionality which
+     * will throttle based on the key expression to group exchanges. This will make key-based throttling
+     * instead of overall throttling.
+     * <p/>
+     * Will default use a time period of 1 second, so setting the maximumRequestCount to eg 10
+     * will default ensure at most 10 messages per second.
+     *
+     * @param maximumRequestCount  an expression to calculate the maximum request count
+     * @param correlationExpressionKey  is a correlation key that can throttle by the given key instead of overall throttling
```
Comment (davsclaus) [inline camel-core/src/main/java/org/apache/camel/model/ProcessorDefinition.java]: The javadoc param should be in the same order as in the method signature

File: camel-core/src/main/java/org/apache/camel/model/ProcessorDefinition.java [hunk 2]
```diff
@@ -2284,6 +2284,48 @@ public ThrottleDefinition throttle(Expression maximumRequestCount) {
...
+     * or that we don't exceed an agreed SLA with some external service.
+     * Here another parameter correlationExpressionKey is introduced for the functionality which
+     * will throttle based on the key expression to group exchanges. This will make key-based throttling
+     * instead of overall throttling.
+     * <p/>
+     * Will default use a time period of 1 second, so setting the maximumRequestCount to eg 10
+     * will default ensure at most 10 messages per second.
+     *
+     * @param maximumRequestCount  an expression to calculate the maximum request count
+     * @param correlationExpressionKey  is a correlation key as an expression that can throttle by the given key instead of overall throttling
```
Comment (davsclaus) [inline camel-core/src/main/java/org/apache/camel/model/ProcessorDefinition.java]: The javadoc param should be in the same order as in the method signature

File: camel-core/src/main/java/org/apache/camel/model/ThrottleDefinition.java [hunk 3]
```diff
@@ -55,14 +56,28 @@
     private Boolean callerRunsWhenRejected;
     @XmlAttribute
     private Boolean rejectExecution;
-    
+    @XmlElement(name = "correlationExpression")
+    private ExpressionSubElementDefinition correlationExpression;
```
Comment (davsclaus) [inline camel-core/src/main/java/org/apache/camel/model/ThrottleDefinition.java]: I would move this up as the correlation expression before the XmlAttribute's

**Comment by onderson** [comment]
for spring dsl, i ll raise another PR. thanks for the reviews

**Comment by onderson** [comment]
just a quick question? i could build camel-spring successfully and i can see regarding changes in target/schema/camel-spring.xsd but still i came across such error running unit tests in camel-spring.

any idea?

`Caused by: org.xml.sax.SAXParseException; lineNumber: 34; columnNumber: 32; cvc-complex-type.2.4.a: Invalid content was found starting with element 'correlationExpression'. One of '{"http://camel.apache.org/schema/spring":description, "http://camel.apache.org/schema/spring":expressionDefinition, "http://camel.apache.org/schema/spring":constant, "http://camel.apache.org/schema/spring":el, "http://camel.apache.org/schema/spring":exchangeProperty, "http://camel.apache.org/schema/spring":groovy, "http://camel.apache.org/schema/spring":header, "http://camel.apache.org/schema/spring":jxpath, "http://camel.apache.org/schema/spring":javaScript, "http://camel.apache.org/schema/spring":jsonpath, "http://camel.apache.org/schema/spring":language, "http://camel.apache.org/schema/spring":method, "http://camel.apache.org/schema/spring":mvel, "http://camel.apache.org/schema/spring":ognl, "http://camel.apache.org/schema/spring":php, "http://camel.apache.org/schema/spring":python, "http://camel.apache.org/schema/spring":ref, "http://camel.apache.org/schema/spring":ruby, "http://camel.apache.org/schema/spring":simple, "http://camel.apache.org/schema/spring":spel, "http://camel.apache.org/schema/spring":sql, "http://camel.apache.org/schema/spring":terser, "http://camel.apache.org/schema/spring":tokenize, "http://camel.apache.org/schema/spring":xtokenize, "http://camel.apache.org/schema/spring":xpath, "http://camel.apache.org/schema/spring":xquery}' is expected.`

**Comment by davsclaus** [comment]
Yeah the XML DSL can be a bit tricky and hence why its good to have unit tests there too. I am in the middle of something and have not updated to latest source. But will try to take a look later - but I am traveling as well this week

**Comment by onderson** [comment]
added similar test case to existing ones. here if you might want to have a look at your convenience.

https://github.com/onderson/camel/tree/CAMEL-6840_spring

as @oscerd suggested, i am trying full build too.. I'll let you know if you don't before me : -)

thank you both.

**Comment by onderson** [comment]
btw, let me tell you here too, that full build did not work for me.

**Comment by WillemJiang** [comment]
I just have a quick look at the code, the schema is updated, but there are something wrong with the xml file.
As the ThrottleDefinition is extends from ExpressionNode, but the first element of ExpressionNode is a reference of expression element, not the correlationExpression that you want.
So you keep getting the complain message.

Once I put the correlationExpression as the last element of throttle, I don't get the error message anymore. (but the test is freeze ). 

```
<route errorHandlerRef="dlc">
  <from uri="seda:b"/>
  <throttle timePeriodMillis="1000">
    <header>max</header>
    <to uri="log:result"/>
    <to uri="mock:result"/>
    <correlationExpression>
      <constant>2</constant>
    </correlationExpression>
  </throttle>
</route>
```

We need to address the element order first to fix this issue.

BTW, I also find an issue in the ThrottleDefinition, to support to old XML which has no correlationExpression, we need to specify the element as an optional one.
Just like this :

```
@XmlElements({
    @XmlElement(name = "correlationExpression", type = ExpressionSubElementDefinition.class)}
)
private ExpressionSubElementDefinition correlationExpression;

```

**Comment by onderson** [comment]
Just reverted the changes in this PR. Implementation is not correct. It is a bit harder than i initially thought of. I will take a look at this again.

**Pull request merged**